river/ensemble/adaptive_random_forest.py

```diff
@@ -0,0 +1,824 @@
+                        index_original=self.index_original,
+                        base_model=self.model.new_instance(),
+                        created_on=n_samples_seen,
+                        base_drift_detector=self.base_drift_detector,
+                        base_warning_detector=self.base_warning_detector,
+                        is_background_learner=True,
+                        base_metric=copy.deepcopy(self.base_metric)
+                    )
+                    # Reset the warning detector for the current object
+                    self.warning_detector = copy.deepcopy(self.base_warning_detector)
+
+            # Update the drift detector
+            self.drift_detector.update(drift_detector_input)
+
+            # Check if there was a change
+            if self.drift_detector.change_detected:
+                self.last_drift_on = n_samples_seen
+                self.n_drifts_detected += 1
+                self.reset(n_samples_seen)
+
+    @abc.abstractmethod
+    def _drift_detector_input(self,
+                              y_true: typing.Union[base.typing.ClfTarget, base.typing.RegTarget],
+                              y_pred: typing.Union[base.typing.ClfTarget, base.typing.RegTarget]):
+        raise NotImplementedError
+
+
+class ForestMemberClassifier(BaseForestMember, base.Classifier):
+    """Forest member class for classification"""
+
+    def __init__(self,
+                 index_original: int,
+                 base_model: ARFHoeffdingTreeClassifier,
+                 created_on: int,
+                 base_drift_detector: base.DriftDetector,
+                 base_warning_detector: base.DriftDetector,
+                 is_background_learner,
+                 base_metric: MultiClassMetric):
+        super().__init__(
+            index_original=index_original,
+            base_model=base_model,
+            created_on=created_on,
+            base_drift_detector=base_drift_detector,
+            base_warning_detector=base_warning_detector,
+            is_background_learner=is_background_learner,
+            base_metric=base_metric
+        )
+
+    def _drift_detector_input(self, y_true: base.typing.ClfTarget, y_pred: base.typing.ClfTarget):
+        return int(not y_true == y_pred)    # Not correctly_classifies
+
+    def predict_one(self, x):
+        return self.model.predict_one(x)
+
+    def predict_proba_one(self, x):
+        return self.model.predict_proba_one(x)
+
+
+class ForestMemberRegressor(BaseForestMember, base.Regressor):
+    """Forest member class for regression"""
+
+    def __init__(self,
+                 index_original: int,
+                 base_model: ARFHoeffdingTreeRegressor,
+                 created_on: int,
+                 base_drift_detector: base.DriftDetector,
+                 base_warning_detector: base.DriftDetector,
+                 is_background_learner,
+                 base_metric: typing.Union[MAE, MSE]):
```

Comment:
According to the docstring in ARFReg, the metric is an instance of any Regression metric. Is it right?

Reply:
I added a note as we have only tested with MAE and MSE.

Reply:
Perfect!